driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -271,11 +267,36 @@ public void onResult(final BsonDocument result, final Throwable t) {
         public void onResult(final QueryResult<T> result, final Throwable t) {
             if (t != null) {
                 connection.release();
+                connectionSource.release();
                 close();
                 callback.onResult(null, t);
             } else {
                 handleGetMoreQueryResult(connection, callback, result);
             }
         }
     }
+
+    private ServerCursor getCursorForNext() {
+        ServerCursor localCursor;
+        synchronized (lock) {
+            localCursor = cursor.get();
+            if (localCursor == null) {
```

Comment:
It maybe the current tests have been historically wrong but simplifying this means some tests will fail as the connectionSource wont be released.